### PR TITLE
Add server TLS certificate verification

### PR DESCRIPTION
**Pull request by takanabe**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes https://github.com/grafana/loki/issues/1986

**Special notes for your reviewer**:

**Checklist**
- [x] Documentation added
- [x] Tests updated

### CODE REVIEW

**Comment by takanabe** [comment]
[Question]

1. Which Ruby version do you use to develop this plugin? I want to push `.ruby-version`.
1. Why don't we include Gemfile.lock in this directory?

File: docs/clients/fluentd/README.md [hunk 1]
```diff
@@ -210,6 +210,21 @@ Specify a pair of client certificate and private key with `cert` and `key` if a
 </match>
 ```
 
+### server certificate verification
+A flag to enable a server certificate verification. By default the `verify_tls` is set to true.
```
Comment (jgehrcke) [inline docs/clients/fluentd/README.md:214]: "to disable server verification"? :)

File: cmd/fluentd/lib/fluent/plugin/out_loki.rb [hunk 2]
```diff
@@ -155,7 +158,7 @@ def ssl_opts(uri)
 
         if !@cert.nil? && !@key.nil?
           opts = opts.merge(
-            verify_mode: OpenSSL::SSL::VERIFY_PEER,
+            verify_mode: @verify_tls ? OpenSSL::SSL::VERIFY_PEER : OpenSSL::SSL::VERIFY_NONE,
```
Comment (jgehrcke) [inline cmd/fluentd/lib/fluent/plugin/out_loki.rb:161]: This code path is about not using a client certificate, the common case.

I think that even when using a client certificate one might want to be able to disable server certificate verification which makes me think that `@verify_tls` should be used in some other code path, too. WDYT?

Comment (jgehrcke) [reply cmd/fluentd/lib/fluent/plugin/out_loki.rb:161]: That is, further above we might want to have something in the lines of
```
        opts = {
          use_ssl: uri.scheme == 'https',
          verify_mode: @verify_tls ? OpenSSL::SSL::VERIFY_PEER : OpenSSL::SSL::VERIFY_NONE,
        }
```

File: docs/clients/fluentd/README.md (hunk 1 above)
Comment (takanabe) [reply docs/clients/fluentd/README.md:214]: This explanation should be fine. The reason why we introduce this flag is to specify server TLS verification mode. Not for disabling the verification. If we think, "to disable" is better, we need to change the option to `no_tls_verify` or something.

File: cmd/fluentd/lib/fluent/plugin/out_loki.rb (hunk 2 above)
Comment (takanabe) [reply cmd/fluentd/lib/fluent/plugin/out_loki.rb]: According to [rdoc](https://ruby-doc.org/stdlib-2.7.1/libdoc/openssl/rdoc/OpenSSL/SSL/SSLContext.html), the default value is `VERIFY_NONE`. 


>Valid modes are VERIFY_NONE, VERIFY_PEER, VERIFY_CLIENT_ONCE, VERIFY_FAIL_IF_NO_PEER_CERT and defined on OpenSSL::SSL, The default mode is VERIFY_NONE, which does not perform any verification at all.


I keep the current default `VERIFY_PEER` but enable developers to turn off the verification with `verify_tls = false` https://github.com/grafana/loki/pull/2171/commits/07e0defa5b19ad894dba840d5402a598c507264e

WDYT?

File: docs/clients/fluentd/README.md (hunk 1 above)
Comment (jgehrcke) [reply docs/clients/fluentd/README.md:214]: I think your thinking around this patch came off the wrong foot. You seem to have thought that we somehow want to be able to _enable_ server certificate verification. It *is* enabled by default and we really look for a way to optionally disable it :).

Comment (jgehrcke) [reply docs/clients/fluentd/README.md:214]: So, let's focus on the actual patch first and then adjust the docs, ok?

**Comment by jgehrcke** [comment]
Some high level feedback, clarifying the motivation and goal.

> According to rdoc, the default value is VERIFY_NONE. 

This is veeeery out-of-date documentation. The Internet and open-source software fortunately moved around 2014 and 2015 :-)

https://github.com/ruby/ruby/blob/ruby_2_7/sample/drb/drbssl_c.rb#L9
https://github.com/ruby/ruby/blob/v2_7_1/lib/rubygems/request.rb#L52

```
    connection.use_ssl = true
    connection.verify_mode =
      Gem.configuration.ssl_verify_mode || OpenSSL::SSL::VERIFY_PEER
```

That is, the default is of course to verify the server certificate when doing HTTPS.

That's how this patch came off the wrong foot. As one can also see when comparing the title of https://github.com/grafana/loki/issues/1986 ("add option to disable server certificate verification") and the title of this PR ("Add server TLS certificate verification").

Really, we need a switch to optionally _disable_ server certificate verification.

> I keep the current default for this plugin (VERIFY_PEER) but enable developers to turn off the verification with verify_tls = false 07e0def

The current patch isn't doing that! In a working patch we will find a part of the (new) code that  sets
```
verify_mode: OpenSSL::SSL::VERIFY_NONE
```
but _only_ when `verify_tls` is set to `false` :-).

Makes sense?

I'll happily review again, just ping me please.

**Comment by takanabe** [comment]
Thank you for the comment! I read the links and [net/http](https://github.com/ruby/ruby/blob/master/lib/net/http.rb#L611) and you are right. The default value of `verify_mode` is `OpenSSL::SSL::VERIFY_PEER`.

I pushed new commit based on your advice https://github.com/grafana/loki/pull/2171/commits/4aadf6079d74f3b655612f9e90a098eb68940534

**Comment by takanabe** [comment]
@jgehrcke could you review the changes?

**Comment by takanabe** [comment]
@cyriltovena thank you for the review!


I pushed the commit to update document https://github.com/grafana/loki/pull/2171/commits/e1dac90ea7b3c9044b171b4355fca0a12fefd291

Should I rebase all commits into one commit or will you squash-merge this PR?

File: docs/clients/fluentd/README.md [hunk 3]
```diff
@@ -210,6 +210,21 @@ Specify a pair of client certificate and private key with `cert` and `key` if a
 </match>
 ```
 
+### server certificate verification
```
Comment (cyriltovena) [inline docs/clients/fluentd/README.md:213]:
```suggestion
### Server certificate verification
```

**Comment by jgehrcke** [comment]
Thanks again everyone here.

**Pull request merged**